hail/src/main/scala/is/hail/expr/ir/agg/TakeByAggregator.scala

```diff
@@ -522,39 +521,33 @@ class TakeByRVAS(val valueVType: VirtualTypeWithReq, val keyVType: VirtualTypeWi
       mb.invokeCode(_, _, _)
     }
 
-    val r = mb.getCodeParam[Region](1)
-    val indicesToSort = mb.newLocal[Long]("indices_to_sort")
-    val i = mb.newLocal[Int]("i")
-    val o = mb.newLocal[Long]("i")
-
-    def indexOffset(idx: Code[Int]): Code[Long] = indicesToSort + idx.toL * 4L
-
-    def indexAt(idx: Code[Int]): Code[Int] = Region.loadInt(indexOffset(idx))
-
-    val srvb = (new StagedRegionValueBuilder(mb, resultType, r))
-    mb.emit(Code(
-      indicesToSort := r.load().allocate(4L, ab.size.toL * 4L),
-      i := 0,
-      Code.whileLoop(i < ab.size,
-        Region.storeInt(indicesToSort + i.toL * 4L, i),
-        i := i + 1),
-      quickSort(indicesToSort, 0, ab.size - 1),
-      srvb.start(ab.size),
-      i := 0,
-      Code.whileLoop(i < ab.size,
-        o := Code.memoize(indexAt(i), "tba_qsort_i") { i => elementOffset(i) },
-        eltTuple.isFieldDefined(o, 1).mux(
-          srvb.addWithDeepCopy(valueType, Region.loadIRIntermediate(valueType)(eltTuple.fieldOffset(o, 1))),
-          srvb.setMissing()
-        ),
-        srvb.advance(),
-        i := i + 1
-      ),
-      srvb.end()
-    ))
+    mb.emitWithBuilder[Long] { cb =>
+      val r = mb.getCodeParam[Region](1)
+
+      val indicesToSort = cb.newLocal[Long]("indices_to_sort",
+        r.load().allocate(4L, ab.size.toL * 4L))
+
+      val i = cb.newLocal[Int]("i", 0)
+
+      cb.whileLoop(i < ab.size, {
+        cb += Region.storeInt(indicesToSort + i.toL * 4L, i)
+        cb.assign(i, i + 1)
+      })
+
+      def indexOffset(idx: Code[Int]): Code[Long] = indicesToSort + idx.toL * 4L
```

Comment:
Move this up, you use this expression above and you can use this function instead.